File: 2020/day08/puzzle8.ts
const input = await Deno.readTextFile("puzzle8.in");

type Instruction = { operation: string; argument: number };
type ExecutionResult = { terminated: boolean; accumulator: number };

function parseInstruction(instruction: string): Instruction {
  const [operation, argument] = instruction.split(" ");
  return { operation, argument: parseInt(argument) };
}

const instructions = input.split("\n").map(parseInstruction);

function swapOperation(instruction: Instruction) {
  if (instruction.operation === "jmp") {
    instruction.operation = "nop";
  } else {
    instruction.operation = "jmp";
  }
}

function executeInstructions(instructions: Instruction[]): ExecutionResult {
  let currentInstruction = 0;
  let accumulator = 0;
  const executedInstructions: number[] = [];

  while (currentInstruction < instructions.length) {
    const { operation, argument } = instructions[currentInstruction];

    if (executedInstructions.includes(currentInstruction)) {
      return { terminated: false, accumulator };
    }

    executedInstructions.push(currentInstruction);
    if (operation === "acc") {
      accumulator += argument;
      currentInstruction += 1;
    } else if (operation === "jmp") {
      currentInstruction += argument;
    } else if (operation === "nop") {
      currentInstruction += 1;
    }
  }

  return { terminated: true, accumulator };
}

function part1() {
  return executeInstructions(instructions).accumulator;
}

function part2() {
  for (const instruction of instructions) {
    const { operation } = instruction;

    if (operation == "acc") {
      continue;
    }

    swapOperation(instruction);
    const result = executeInstructions(instructions);
    if (result.terminated) {
      return result.accumulator;
    }
    swapOperation(instruction);
  }
}

console.log(part1());
console.log(part2());

File: 2020/day08/puzzle8.py
file = open('puzzle8.in')

instructions = []
executed_instuctions = []

for line in file:
  [operation, argument] = line.strip().split(' ')
  instructions.append({ 'operation': operation, 'argument': int(argument) })

def part1():
  accumulator = 0
  current_instruction = 0

  while True:
    if current_instruction in executed_instuctions:
      break

    executed_instuctions.append(current_instruction)

    if instructions[current_instruction]['operation'] == 'nop':
      current_instruction += 1
      continue

    if instructions[current_instruction]['operation'] == 'jmp':
      current_instruction += instructions[current_instruction]['argument']
    elif instructions[current_instruction]['operation'] == 'acc':
      accumulator += instructions[current_instruction]['argument']
      current_instruction += 1

  return accumulator

print(part1())

executed_instuctions = []
swapped_instructions = []

def swap_instruction(instruction_index):
  #print('swapping', instruction_index+1, 'from', instructions[instruction_index]['operation'])
  if instructions[instruction_index]['operation'] == 'jmp':
    instructions[instruction_index]['operation'] = 'nop'
  else:
    instructions[instruction_index]['operation'] = 'jmp'
  
def go_back():
  accumulator = 0
  if len(swapped_instructions) > 0:
    swapped_instruction = swapped_instructions[-1]
  else:
    swapped_instruction = None

  while True:
    instruction_to_revert = executed_instuctions.pop()
    #print('going back from', current_instruction + 1, 'to', instruction_to_revert + 1, accumulator)

    if instructions[instruction_to_revert]['operation'] == 'acc':
      accumulator -= instructions[instruction_to_revert]['argument']
    else:      
      if instruction_to_revert == swapped_instruction:
        swap_instruction(instruction_to_revert)
        swapped_instruction = None

      if swapped_instruction == None and not instruction_to_revert in swapped_instructions:
        swapped_instructions.append(instruction_to_revert)
        swap_instruction(instruction_to_revert)
        break
  
  return [instruction_to_revert, accumulator]

def part2():
  accumulator = 0
  current_instruction = 0

  while current_instruction < len(instructions):
    if current_instruction in executed_instuctions:
      [new_instruction, accumulator_change] = go_back()

      current_instruction = new_instruction
      accumulator += accumulator_change

    #print('forward', current_instruction + 1, instructions[current_instruction], accumulator)
    executed_instuctions.append(current_instruction)

    if instructions[current_instruction]['operation'] == 'acc':
      accumulator += instructions[current_instruction]['argument']
      current_instruction += 1
    elif instructions[current_instruction]['operation'] == 'nop':
      current_instruction += 1
    elif instructions[current_instruction]['operation'] == 'jmp':
      current_instruction += instructions[current_instruction]['argument']

  return accumulator

print(part2())
